### DISCUSSION TITLE
V7, Watch with typescript

### DISCUSSION THREAD

**Discussion by michael-land** [react-hook-form/react-hook-form#4305] [Q&A]
**original problem**

in v7, the watch return array instead of object, how could we keep watched value typesafe?

https://codesandbox.io/s/material-demo-forked-pifom?file=/demo.tsx

```ts
import { useForm, useWatch } from "react-hook-form";

interface Schema {
  a: number;
  b: string;
  c: boolean;
}
export default function Demo() {
  const form = useForm<Schema>();

  // infer to a: string | number | boolean
  // infer to b: string | number | boolean
  // infer to c: string | number | boolean

  const [a, b, c] = useWatch<Schema>({
    name: ["a", "b", "c"],
    control: form.control
  });

  return null;
}
```

And if any value is defined as any, then all value widen to any

```ts
import { useForm, useWatch } from "react-hook-form";

interface Schema {
  a: number;
  b: string;
  c: boolean;
  d: any
}
export default function Demo() {
  const form = useForm<Schema>();

  // infer to a: any
  // infer to b: any
  // infer to c: any
  // infer to d: any

  const [a, b, c] = useWatch<Schema>({
    name: ["a", "b", "c", "d"],
    control: form.control
  });

  return null;
}
```

**Updated**

I think typesafe is really important for form library, it's also the goal for v7 which declare to support typescript better. If we're unable to make tuple typesafe, I suggest we revert to object.

Here is an example using `type-fest` ( just a quick demo to show it works, we can write our own helper )

https://codesandbox.io/s/zealous-wing-i8wex?file=/src/index.ts

```ts
import { Get } from "type-fest";

function pick<Schema, Path extends string>(
  object: Schema,
  names: Path[]
): { [K in Path]: Get<Schema, K> } {
  return {} as any;
}

const obj = {} as {
  a: number;
  b: string;
  c: boolean;
  d: any;
  array: number[];
  objectArray: { a: number }[];
  nestObjectArray: { a: { b: number }[] }[];
};

const a = pick(obj, [
  "unknown_key",
  "a",
  "b",
  "objectArray.0",
  "objectArray.0.a",
  "nestObjectArray.0.a.0.b"
]);


// I get 
const a: {
    a: number;
    b: string;
    unknown_key: unknown;
    "objectArray.0": {
        a: number;
    };
    "objectArray.0.a": number;
    "nestObjectArray.0.a.0.b": number;
}
```

**Comment by bluebill1049** [comment]
If you are requesting an array of watched values, it makes sense to return an array of them. I would seek help to make it work rather than revert the change.

**Reply by michael-land** [reply]
put my two cents in.

In the js ecosystem, it common to return object when requesting an array of fields, e.g. [lodash.pick](https://lodash.com/docs/4.17.15#pick), [ramda.pick](https://ramdajs.com/docs/#pick) 

It make sense to return array of watched values if we can solve the type issue, :-), otherwise, we lose build time error.

for example

```ts
const [subtotal, tax, discount, shipping] = watch(['subtotal', 'tax', 'discount', 'shipping', ...and other pricing])

const total = subtotal + tax - discount + shipping + ...other pricing;

// imagine the you remove discount from watched fields but forgot to alter the return tuple. 
// You won't get any type error or run time error, which is hard to debug

//  the discount field was removed from watched array
const [subtotal, tax, discount, shipping] = watch(['subtotal', 'tax', 'shipping', ...and other pricing])

const total = subtotal + tax - discount + shipping + ...other pricing;
```

**Reply by bluebill1049** [reply]
valid points. I would say let's try to resolve it first: 

> It make sense to return array of watched values if we can solve the type issue, :-), otherwise, we lose build time error.

**Reply by bluebill1049** [reply]
looks like its a problem of inferring the type: https://codesandbox.io/s/material-demo-forked-pifom?file=/demo.tsx

**Reply by bluebill1049** [reply]
if you have time, please take a look at useWatch.ts see if you can spot anything.

**Reply by michael-land** [reply]
I tried to solve it but no luck. not a type expert. =.=

**Reply by bluebill1049** [reply]
you saw this works correctly for the type right: 

const [a, b, c] = useWatch<Schema, ['a','b', 'c']>({
    name: ["a", "b", "c"],
    control: form.control
  });

**Reply by michael-land** [reply]
Umm, I just noticed that, that's interesting. Looks like the problem is the type `FieldPath<>`?

**Reply by bluebill1049** [reply]
@leapful any chance u have some idea?

**Reply by bluebill1049** [reply]
This problem exisits in `watch` API as well.

**Reply by bluebill1049** [reply]
cc @kotarella1110 as well.

**Reply by bluebill1049** [reply]
problem due to this: 

`TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>` which will return all the field input string name. how can i extra the second argument is what i am missing here.

**Reply by bluebill1049** [reply]
If this is truly not possible, we will have to include that in the doc.  as watch or useWatch will need to declare on the second generic.

**Reply by michael-land** [reply]
I’m guessing that the only way to go. 

You don’t want go back to object right? Which has better type support. 

Maybe return array in the next major version where typescript might handle tuple type better.

it’s cumbersome to write them twice :-/

**Reply by michael-land** [reply]
I’m totally ok with array result. I can wrap useWatch in my own hook.

Just want to share my thoughts

**Reply by bluebill1049** [reply]
> You don’t want go back to object right? Which has better type support.

I am happy to go back, but that wouldn't solve the problem. because `name` is getting type-checked which is results in the write again. Let's see if some genius can come up with a solution.

**Comment by Mendoza920** [comment]
I need help with sandbox
I dont knkw how to use it

**Reply by michael-land** [reply]
here are some templates 
- [useForm](https://codesandbox.io/s/react-hook-form-useform-template-lsm61) Template
- [useFieldArray](https://codesandbox.io/s/react-hook-form-usefieldarray-template-q0jek) Template
- [Controller](https://codesandbox.io/s/react-hook-form-controller-template-yu0z5) Template

**Comment by Mendoza920** [comment]
I need physical help

**Comment by leapful** [comment]
@bluebill1049 

When using "useWatch()" without second type definition for FieldPath array, the default type FieldPath<Schema>[] will be used as type for prop "name", something like 
```
ValueOf<{ a: number, b: string, c: boolean }>[].
```
or
```
("a" | "b" | "c")[]
```
Therefore it does not know how to set type for each item of the return array because there is no order.

When using "useWatch<Schema, ["a", "b", "c"]>()", the order is now set that's why it can show the correct return type value.

I think we need to specify the type for prop "name" in this case, otherwise, we need to find the way to infer order of array field path.

**Reply by bluebill1049** [reply]
Thanks, @leapful for the investigation. I was hoping for a way to infer, but it hit my ts knowledge limitation...

**Comment by lop862** [comment]
Screencast

**Reply by lop862** [reply]
Pantalla inlabrica
